src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -1103,41 +1099,29 @@ private class Emit(
 
       case If(cond, cnsq, altr) =>
         val codeCond = emit(cond)
-        val xmcond = mb.newLocal[Boolean]()
-        val xvcond = mb.newLocal[Boolean]()
         val mout = mb.newLocal[Boolean]()
-        val aCnsq = emitArrayIterator(cnsq)
-        val aAltr = emitArrayIterator(altr)
+        val cnsqArray = emitArrayIterator(cnsq)
+        val altrArray = emitArrayIterator(altr)
 
         val f = { cont: F =>
-          val addCnsq = aCnsq.arrayEmitter(cont)
-          val addAltr = aAltr.arrayEmitter(cont)
+          def addElements(array: ArrayIteratorTriplet): Code[Unit] = {
+            val emitter = array.arrayEmitter(cont)
+            Code(
+              emitter.setup,
+              mout := emitter.m.getOrElse(false),
+              mout.mux(Code._empty, Code(array.calcLength, emitter.addElements)))
+          }
+
           val setup = Code(
             codeCond.setup,
-            xmcond := codeCond.m,
-            xmcond.mux(
-              Code(
-                mout := true,
-                xvcond := coerce[Boolean](defaultValue(cond.typ))
-              ),
-              Code(
-                mout := false,
-                xvcond := coerce[Boolean](codeCond.v),
-                addCnsq.setup, addAltr.setup)))
-          val missing: Code[Boolean] = if (addCnsq.m.isEmpty && addAltr.m.isEmpty)
-            mout
-          else
-            Code(
-              xmcond.mux(Code._empty,
-                mout := xvcond.mux(addCnsq.m.getOrElse(false), addAltr.m.getOrElse(false))),
-              mout)
-          val add = xvcond.mux(addCnsq.addElements, addAltr.addElements)
-          EmitArrayTriplet(setup, Some(missing), add)
+            codeCond.m.mux(
+              mout := true,
+              coerce[Boolean](codeCond.v).mux(
+                addElements(cnsqArray), addElements(altrArray))))
+          EmitArrayTriplet(setup, Some(mout.load()), Code._empty)
         }
 
-        val lenCalc = xvcond.mux(aCnsq.calcLength, aAltr.calcLength)
-        val optLen = aCnsq.length.flatMap(l1 => aAltr.length.map(xvcond.mux(l1, _)))
-        ArrayIteratorTriplet(lenCalc, optLen, f)
+        ArrayIteratorTriplet(Code._empty, None, f)
```

Comment:
this solution is much simpler and I like it, but it does look like we pay by obscuring the length through an `if`. Do I understand that correctly?

Reply:
Yeah.

Reply:
I think I might be able to keep the length stuff by storing basically everything in the `emitArrayIterator` passes in fields instead of locals, which mostly just sidesteps the verifier issue because the fields always have default values. I'm actually kind of leaning towards that now, what do you think?

Reply:
ugh, actually this is kind of annoying because i'd either have to do this for like everything in Emit, or stash things in fields when they pop into the ArrayIterator logic. I can try to do that, if you like.

Reply:
Let's leave it for another day.